mdn/src/main/java/org/apache/james/mdn/MDNReportParser.java

```diff
@@ -292,8 +288,42 @@ Rule dispositionNotificationContent() {
         /*    reporting-ua-field = "Reporting-UA" ":" OWS ua-name OWS [
                                    ";" OWS ua-product OWS ]    */
         Rule reportingUaField() {
-            return Sequence("Reporting-UA", ":", ows(), uaName(), ows(),
-                Optional(Sequence(";", ows(), uaProduct(), ows())));
+            return Sequence(
+                push(ReportingUserAgent.builder()),
+                "Reporting-UA", ":", ows(), uaName(), ACTION(setUserAgentName()), ows(),
+                Optional(Sequence(";", ows(), uaProduct(), ACTION(setUserAgentProduct()), ows())),
+                ACTION(buildReportingUserAgent())
+                );
+        }
+
+        boolean buildReportingUserAgent() {
+            push(this.<ReportingUserAgent.Builder>popT().build());
+            return true;
+        }
+
+        boolean setUserAgentName() {
+            this.<ReportingUserAgent.Builder>peekT().userAgentName(match());
+            return true;
+        }
+
+        boolean setUserAgentProduct() {
+            this.<ReportingUserAgent.Builder>peekT().userAgentProduct(match());
+            return true;
+        }
+
+        @SuppressWarnings("unchecked")
+        <T> T popT() {
```

Comment:
I would create a class with these utility methods because it's not related to our parser.
Also, I would try to get rid of the painful `return true` by providing a wrapper that, given a lambda, would call `ACTION` and do the `return true` for us.
Anyway, feel free to merge this PR like that, it can be done later.

Reply:
I extracted the methods in a separated (and inherited from it as it is needed for parboiled...).
For the rest, it seems to not be doable at first glance with parboiled, as ACTION() is used as a marker for enhancing the code at runtime.